### DISCUSSION TITLE
[DragonBones] How to replaceDisplay with external image?

### DISCUSSION THREAD

**Discussion by yunlinos** [korlibs/korge#488] [Q&A]
Hello,
I want change slot display with method: ` fun replaceDisplay(slot: Slot, displayData: DisplayData?, displayIndex: Int = -1)` , but don‘t known how to generate external image to **ImageDiaplayData**?

is it possible to change slot display with external image? If so, how can I do it?

**Comment by soywiz** [comment]
@yunlinos the API is a port of the typescript one and I don't remember the details. Looking at the sample, it seems that there is a replaceSkin:
https://github.com/korlibs/korge-samples/blob/5a259513363945d6edcc5909154e43b484b0b89e/samples/dragonbones/src/commonMain/kotlin/main.kt#L475

Does that work for you, or do you need the replaceDisplay function?

**Comment by yunlinos** [comment]
> 
Sorry, it not work for me, the replaceDisplay function is already in **com.dragonbones.factory.BaseFactory**, but I don't know how to create DisplayData from an image.

I have seen the sample, and the replaceSkin method works fine, sadly it needs to parse the DragonBones data ahead of time, such as [ske.json, tex.json,text.png].

My functional scene is an application that can change clothes, there are more than a dozen hats that can be replaced, and these hats are separate pictures, and the size is the same, the animation is also the same, so I want to directly replace the display of the slot through the external picture .

So my requirement is: to directly parse a picture as the display of the slot, there are similar methods in the official demo, for example: http://developer.egret.com/#160-db-change-disp

```
    private c1 = ["m_121005_F_S_png","m_121005_F_L1_png","m_121005_F_L2_png","m_121005_F_R1_png","m_121005_F_R2_png"];
    private c2 = ["m_121011_F_S_png","m_121011_F_L1_png","m_121011_F_L2_png","m_121011_F_R1_png","m_121011_F_R2_png"];

    //设置上衣
    //参数为即将替换的纹理集ID
    public setCloth( val:string[] ):void
    {
        this.setNewSlot( "Ayifu"    , val[0] );
        this.setNewSlot( "Axiuzi11" , val[1] );
        this.setNewSlot( "Axiuzi21" , val[2] );
        this.setNewSlot( "Axiuzi12" , val[3] );
        this.setNewSlot( "Axiuzi22" , val[4] );
    }

    //针对slot设置其新内容
    private setNewSlot( slotName:string, textureName:string )
    {
        //方法1
        var slot:dragonBones.Slot = this._armature.getSlot( slotName );
        var b:egret.Bitmap = new egret.Bitmap();
        b.texture = RES.getRes( textureName );
        b.x = slot.display.x;
        b.y = slot.display.y;
        b.anchorOffsetX = b.width/2;
        b.anchorOffsetY = b.height/2;
        slot.setDisplay( b );

        //方法2，仅限于slot中内容为Bitmap
        //var slot:dragonBones.Slot = this._armature.getSlot(slotName);
        //slot.display.texture = RES.getRes(textureName);
    }
```

**Comment by yunlinos** [comment]
@soywiz Looking forward to your reply，thanks

**Comment by soywiz** [comment]
I have been checking this, and I'm not fully sure. The method you want to use is marked as `@private` in a comment. Maybe typescript didn't have visibility information, but that method shouldn't be available.
In the snipet you put, it uses the setDisplay method that is marked as deprecated: https://github.com/korlibs/korge/blob/1fe2c8a34949b57f5971346b0443549dc6e310af/korge-dragonbones/src/commonMain/kotlin/com/dragonbones/armature/Slot.kt#L1359
https://github.com/korlibs/korge/blob/1fe2c8a34949b57f5971346b0443549dc6e310af/korge-dragonbones/src/commonMain/kotlin/com/dragonbones/armature/Slot.kt#L51
Not sure what's the officla way for doing it after the deprecation, since comments are in chinese.
It accepts anything, so I'm not sure what kind of object would it accept. You can try a `Bitmap` or a `Image(bitmap)` and see if it helps?
Like:
```kotlin
slot.display = bitmap
slot.display = bitmap.slice()
slot.display = Image(bitmap)
```
Or the like?

EDIT:
It seems it should be a view, so `Image(bitmap)` might work.

https://github.com/korlibs/korge/blob/1fe2c8a34949b57f5971346b0443549dc6e310af/korge-dragonbones/src/commonMain/kotlin/com/soywiz/korge/dragonbones/KorgeDbSlot.kt#L73

**Reply by yunlinos** [reply]
Yes, **Image(bitmap)** is work. However, through this method , the position and scale of the display relative to the slot are not easy to control. 
https://docs.egret.com/dragonbones/cn/docs/dbLibs/5foramt
```json
  "transform": {
      "x": 0.00, // 水平位移 (可选属性 默认: 0.00)
      "y": 0.00, // y 垂直位移 (可选属性 默认: 0.00)
      "skX": 0.0000, // 水平斜切 (可选属性 默认: 0.0000)
      "skY": 0.0000, // 垂直斜切 (可选属性 默认: 0.0000)
      "scX": 1.0000, // 垂直缩放 (可选属性 默认: 1.0000)
      "scY": 1.0000, // 垂直缩放 (可选属性 默认: 1.0000)
  },
```

**Reply by yunlinos** [reply]
I try change the transform, but it not work.
```kotlin
onUpAnywhere {
    val slot = armatureDisplay.armature.getSlot("weapon_hand_l") ?: return@onUpAnywhere
    val bitmap = resourcesVfs["icon.png"].readBitmap()
    slot.display = Image(bitmap).apply {
      // this is not work
        x = 91.22
        y = -30.21
    }
}
```

**Comment by soywiz** [comment]
@yunlinos I will have to investigate further, because I only migrated the typescript code and the samples, and it was sometime ago, so I don't remember the details. But will be nice to have a sample covering that keeping the scale as expected.
I have created this issue: https://github.com/korlibs/korge/issues/492 I will try to check asap, but next days I'm going to be busy so not sure when I will be able to check it

**Reply by yunlinos** [reply]
OK, thanks. @soywiz 

`slot.display = Image(bitmap) ` is a very simple way to modify the display, it would be better if we can modify the Transform information. In order to temporarily solve this problem, I used another solution that is not very friendly, which is to replace the textureData of the slot, and at the same time, i can set the corresponding Transform information.

```kotlin
protected suspend fun Slot.changeSlotDisplay(replaceBitmap: Bitmap, transformDb: TransformDb? = null) {
    // use bitmap HasCode as DragonBonesName and ArmatureName
    val replaceArmatureName = replaceBitmap.hashCode().toString()
    // displayName of slot
    val replaceDisplayName = this._displayFrames.first { it.rawDisplayData != null }.rawDisplayData!!.name
    
    val mockTexModel = mockTexModel(replaceArmatureName, replaceDisplayName, replaceBitmap.width, replaceBitmap.height)
    val textureAtlasData = withContext(Dispatchers.IO) {
        Json.parse(gson.toJson(mockTexModel))!!
    }
    // textureAtlasData = {height=240, width=240, imagePath="", name=24093607, SubTexture=[{frameHeight=-1, frameWidth=-1, frameX=0, frameY=0, height=120, name=weapon_1004_l, width=212, x=0, y=0}]}
    factory.parseTextureAtlasData(textureAtlasData, replaceBitmap.mipmaps())

    // find and replace textureData of display
    val replaceTextureData = getReplaceDisplayTextureData(replaceArmatureName, replaceDisplayName)
    this.replaceTextureData(replaceTextureData)

    if (transformDb != null) {
        this._displayFrame?.rawDisplayData?.transform?.copyFrom(transformDb)
    }
}
```
